File: src/forge/core/models/adapter.py
"""Adapter models — capability declarations and lifecycle state.

Every external system connects to Forge through an adapter.
Adapters are plugins that conform to a standard interface
governed by FACTS specs.
"""

from __future__ import annotations

from datetime import datetime  # noqa: TC003 — Pydantic needs runtime access
from forge._compat import StrEnum
from typing import Any

from pydantic import BaseModel, Field


class AdapterTier(StrEnum):
    """The system tier this adapter connects to."""

    OT = "OT"  # PLC, SCADA, HMI, DCS
    MES_MOM = "MES_MOM"  # MES, QMS, WMS, LIMS, CMMS, EBR
    ERP_BUSINESS = "ERP_BUSINESS"  # ERP, SCM, CRM, PLM, BI
    HISTORIAN = "HISTORIAN"  # OSIsoft PI, AVEVA, etc.
    DOCUMENT = "DOCUMENT"  # SharePoint, file shares, DMS


class AdapterState(StrEnum):
    """Adapter lifecycle states."""

    REGISTERED = "REGISTERED"
    CONNECTING = "CONNECTING"
    HEALTHY = "HEALTHY"
    DEGRADED = "DEGRADED"
    FAILED = "FAILED"
    STOPPED = "STOPPED"


class AdapterCapabilities(BaseModel):
    """What this adapter can do."""

    read: bool = True
    write: bool = False
    subscribe: bool = False
    backfill: bool = False
    discover: bool = False


class ConnectionParam(BaseModel):
    """A parameter required to connect to the source system."""

    name: str
    description: str | None = None
    required: bool = True
    secret: bool = False
    default: str | None = None


class DataContract(BaseModel):
    """What this adapter produces."""

    schema_ref: str
    output_format: str = "contextual_record"
    context_fields: list[str] = Field(default_factory=list)


class AdapterManifest(BaseModel):
    """Declares an adapter's identity, capabilities, and data contract.

    The manifest is the adapter's self-description. It is loaded at
    registration time and validated against the FACTS schema. The hub
    knows nothing about adapter internals — only what the manifest declares.
    """

    adapter_id: str
    name: str
    version: str
    type: str = "INGESTION"
    protocol: str
    tier: AdapterTier
    capabilities: AdapterCapabilities = Field(default_factory=AdapterCapabilities)
    data_contract: DataContract
    health_check_interval_ms: int = 5000
    connection_params: list[ConnectionParam] = Field(default_factory=list)
    auth_methods: list[str] = Field(default_factory=lambda: ["none"])
    metadata: dict[str, Any] = Field(default_factory=dict)


class AdapterHealth(BaseModel):
    """Current health status of an adapter instance."""

    adapter_id: str
    state: AdapterState
    last_check: datetime | None = None
    last_healthy: datetime | None = None
    error_message: str | None = None
    records_collected: int = 0
    records_failed: int = 0
    uptime_seconds: float = 0.0

File: go/internal/models/adapter.go
package models

import "time"

// AdapterTier indicates which system tier an adapter connects to.
type AdapterTier string

const (
	AdapterTierOT          AdapterTier = "OT"
	AdapterTierMESMOM      AdapterTier = "MES_MOM"
	AdapterTierERPBusiness AdapterTier = "ERP_BUSINESS"
	AdapterTierHistorian   AdapterTier = "HISTORIAN"
	AdapterTierDocument    AdapterTier = "DOCUMENT"
)

// AdapterState represents the adapter lifecycle state machine.
type AdapterState string

const (
	AdapterStateRegistered AdapterState = "REGISTERED"
	AdapterStateConnecting AdapterState = "CONNECTING"
	AdapterStateHealthy    AdapterState = "HEALTHY"
	AdapterStateDegraded   AdapterState = "DEGRADED"
	AdapterStateFailed     AdapterState = "FAILED"
	AdapterStateStopped    AdapterState = "STOPPED"
)

// AdapterCapabilities declares what an adapter can do.
type AdapterCapabilities struct {
	Read      bool `json:"read"`
	Write     bool `json:"write"`
	Subscribe bool `json:"subscribe"`
	Backfill  bool `json:"backfill"`
	Discover  bool `json:"discover"`
}

// DefaultAdapterCapabilities returns capabilities with read=true, all others false.
func DefaultAdapterCapabilities() AdapterCapabilities {
	return AdapterCapabilities{Read: true}
}

// ConnectionParam describes a parameter required to connect to the source system.
type ConnectionParam struct {
	Name        string `json:"name"`
	Description string `json:"description,omitempty"`
	Required    bool   `json:"required"`
	Secret      bool   `json:"secret"`
	Default     string `json:"default,omitempty"`
}

// DataContract declares what an adapter produces.
type DataContract struct {
	SchemaRef     string   `json:"schema_ref"`
	OutputFormat  string   `json:"output_format"`
	ContextFields []string `json:"context_fields,omitempty"`
}

// AdapterManifest is the adapter's self-description document.
// It is validated against the FACTS schema at registration time.
type AdapterManifest struct {
	AdapterID            string              `json:"adapter_id"`
	Name                 string              `json:"name"`
	Version              string              `json:"version"`
	Type                 string              `json:"type"`
	Protocol             string              `json:"protocol"`
	Tier                 AdapterTier         `json:"tier"`
	Capabilities         AdapterCapabilities `json:"capabilities"`
	DataContract         DataContract        `json:"data_contract"`
	HealthCheckIntervalMs int                `json:"health_check_interval_ms"`
	ConnectionParams     []ConnectionParam   `json:"connection_params,omitempty"`
	AuthMethods          []string            `json:"auth_methods,omitempty"`
	Metadata             map[string]any      `json:"metadata,omitempty"`
}

// AdapterHealth holds the current health status of an adapter instance.
type AdapterHealth struct {
	AdapterID        string       `json:"adapter_id"`
	State            AdapterState `json:"state"`
	LastCheck        *time.Time   `json:"last_check,omitempty"`
	LastHealthy      *time.Time   `json:"last_healthy,omitempty"`
	ErrorMessage     string       `json:"error_message,omitempty"`
	RecordsCollected int64        `json:"records_collected"`
	RecordsFailed    int64        `json:"records_failed"`
	UptimeSeconds    float64      `json:"uptime_seconds"`
}
